Marlin/src/module/planner.cpp

```diff
@@ -794,22 +794,27 @@ block_t* Planner::get_current_block() {
  */
 void Planner::calculate_trapezoid_for_block(block_t * const block, const_float_t entry_factor, const_float_t exit_factor) {
```

Comment:
I've been playing with unit testing this function. I put a few basic tests together, and after merging your changes into it the behavior changes were minimal, which was probably expected.

Do any of the changes in `calculate_trapezoid_for_block` fix major issues, or were they all just minor "off by one" errors in the acceleration/deceleration timing?

It seems like the more impactful changes in this PR are outside of this function?

Reply:
The "off by one" issues result in acceleration spikes that I wouldn't call minor but indeed that's nearly all the changes. I'd call out these cases to unit test:
- an almost almost full acceleration and an almost almost full deceleration segment
- when due to rounding the result would be `cruise_rate` being higher than `nominal_rate`

Reply:
OK, I'll give that a try. I know I have seen cruise being slightly higher than nominal in my tests, although I think that was true even after your changes. I need to look back over what I did to make sure I did things properly.

I also tried replicating the problems you're fixing in the simulator, but without a lot of success. I might be mixing some things up as I juggle branches and tools.

Reply:
I'm unfortunately not going to have time to look back at this until next week sometime.